### PR TITLE
Add WaitFor{Low,High} for stm32l0 exti pins

### PR DESCRIPTION
**Pull request by michaelbeaumont**
Still needs to be tested

### CODE REVIEW

**Comment by Dirbaio** [comment]
You might want to rebase on top of #82, which removed ExtiManager, simplifying the API :) 

I think the logic could be simplified a bit. In wait_for_high, the pin is either:
- initially high, in which case we can return immediately, or
- initially low, in which case we just have to wait for a rising edge with InterruptFuture.

I think that would allow not having a `WaitForHighFuture`? Same for wait_for_low.

**Comment by Dirbaio** [comment]
This could have a race condition. For example, in wait_for_high:

- we read the pin, see it's low
- right after, the pin becomes high
- we setup the exti interrupt for rising edge and wait for it

in that case the rising edge would get missed.

The solution would be FIRST setting up the exti, and THEN checking if it's already high. I can't think of any possible race with that

**Comment by michaelbeaumont** [comment]
> This could have a race condition. For example, in wait_for_high

> The solution would be FIRST setting up the exti, and THEN checking if it's already high. I can't think of any possible race with that

Yep, that's what I'm doing with `WaitForXFuture` I think?

**Comment by Dirbaio** [comment]
> Yep, that's what I'm doing with WaitForXFuture I think?

Yes, but you're checking for high at every poll. It's only necessary to check at the very beginning in the `wait_for_high`.

**Comment by michaelbeaumont** [comment]
> > Yep, that's what I'm doing with WaitForXFuture I think?
> 
> Yes, but you're checking for high at every poll. It's only necessary to check at the very beginning in the `wait_for_high`.

I was under the impression that setting up `exti.listen_gpio` and creating `InterruptFuture` but not yet polling it wouldn't catch the interrupt in the race condition case, but may be wrong!

**Comment by Dirbaio** [comment]
InterruptFuture is designed so it starts "listening" for interrupts from the moment you create it (not from the moment you first poll it). Precisely to help with these kinds of races :)

**Comment by michaelbeaumont** [comment]
I had this exact implementation already, for https://github.com/michaelbeaumont/dht-sensor/pull/14, but was struggling with hanging and flakiness and suspected it might be due to a race condition here but may be more likely due to the `Delay` implementation.

File: embassy-stm32l0/src/exti.rs [hunk 1]
```diff
@@ -51,6 +54,47 @@ impl<T: PinWithInterrupt + 'static> ExtiPin<T> {
     }
 }
 
+impl<T: InputPin + PinWithInterrupt + 'static> ExtiPin<T> {
+    fn wait_for_state<'a>(self: Pin<&'a mut Self>, state: bool) -> impl Future<Output = ()> + 'a {
+        let line = self.pin.line();
+        let s = unsafe { self.get_unchecked_mut() };
+
+        Exti::unpend(line);
```
Comment (Dirbaio) [inline embassy-stm32l0/src/exti.rs:62]: the InterruptFuture already does unpend of the interupt, I think this is not needed.

File: embassy-stm32l0/src/exti.rs [hunk 2]
```diff
@@ -51,6 +54,47 @@ impl<T: PinWithInterrupt + 'static> ExtiPin<T> {
...
+                s.pin.is_high()
+            } else {
+                s.pin.is_low()
+            }
+            .unwrap_or(false);
+            if pin_has_state {
+                return ();
+            }
+
+            fut.await;
```
Comment (Dirbaio) [inline embassy-stm32l0/src/exti.rs:91]: Same.

**Review by Dirbaio — CHANGES_REQUESTED** [review]

File: embassy-stm32l0/src/exti.rs (hunk 1 above)
Comment (xoviat) [reply embassy-stm32l0/src/exti.rs:62]: This unpends the exti line, not the NVIC.

Comment (xoviat) [reply embassy-stm32l0/src/exti.rs:62]: see https://github.com/embassy-rs/embassy/blob/master/embassy-stm32f4/src/exti.rs#L108

Comment (Dirbaio) [reply embassy-stm32l0/src/exti.rs:62]: oh whoops my bad :)

**Pull request merged**